### DISCUSSION TITLE
Follow Composer's SemVer for pre-release versions, please

### DISCUSSION THREAD

**Discussion by simonhamp** [laravel/framework#53158] [Q&A]
@jessarcher a couple of times now I've bumped into Prompts pre-releases breaking NativePHP thanks to the minor version number increase.

Please note Composer's own guidance on this as identified in the [version constraints summary](https://getcomposer.org/doc/articles/versions.md#summary).

If the major version is `0`, when using the `^` constraint, it's treated as the major version. So if I pin to `0.1` and you release `0.2`, I won't be able to retrieve it unless I change the constraint.

This is by design so that I don't end up with breaking changes in my app, while Prompts is not yet "stable".

I'm using the constraint as intended, but as `laravel/laravel` is updating to require the newly minted version of Prompts, it's breaking installations of NativePHP unless I [update](https://github.com/NativePHP/electron/commit/d33e94771523b2de78f28e2e6ac2d520bc97b3aa) the [constraint](https://github.com/NativePHP/electron/pull/110) in the composer.json.

That's fine if there were actual breaking changes, but from my reading of the changes in both the [`0.2`](https://github.com/laravel/prompts/releases/tag/v0.2.0) and [`0.3`](https://github.com/laravel/prompts/releases/tag/v0.3.0) releases there haven't been any breaking changes, so I'm struggling to understand the minor ("major") version bump in these cases.

It feels like I could broaden the constraint on my end (e.g. to something like `~0.3.0` - not sure yet if this would work tho), but then I risk getting bitten by a breaking change.

Please can you clarify here what your approach is so I can determine the appropriate constraints to use?

**Comment by crynobone** [comment]
`0.2.0` was probably shouldn't have been a breaking change release but `0.3.0` is. Moving `illuminate/collections` from `require` to `require-dev` may seems to be harmless but it can potentially cause breaking changes for projects outside of Laravel that's might have expected `illuminate/collections` always to be available.

**Reply by simonhamp** [reply]
I understand that but I don't believe you should consider third-party dependents this way... if a project was relying on `illuminate/collections`, it should be declared in their own `composer.json`.

If they fail to do so, that's on them.

**Reply by crynobone** [reply]
https://github.com/NativePHP/electron/blob/8002c116dc7ed1a942af2fd7caab5594743355e3/composer.json#L34-L39

Using the same logic, this project should also require `illuminate/console` and `illuminate/support`

**Reply by simonhamp** [reply]
Absolutely! Thanks for pointing that out 🙏🏼 I haven't fully audited the dependencies here yet.
